pkg/api/pool_test.go

```diff
@@ -0,0 +1,192 @@
+func (testManager) Description() string                                         { return "deny checker aspect manager for testing" }
+
+func (m testManager) NewAspect(cfg *aspect.CombinedConfig, adapter adapter.Builder, env adapter.Env) (aspect.Wrapper, error) {
+	return m.instance, nil
+}
+func (m testManager) NewDenyChecker(env adapter.Env, c adapter.AspectConfig) (adapter.DenyCheckerAspect, error) {
+	return m.instance, nil
+}
+
+type testAspect struct {
+	fn testAspectFn
+}
+
+func (testAspect) Close() error { return nil }
+func (t testAspect) Execute(attrs attribute.Bag, mapper expr.Evaluator) (*aspect.Output, error) {
+	return t.fn()
+}
+func (testAspect) Deny() status.Status { return status.Status{Code: int32(code.Code_INTERNAL)} }
+
+var (
+	name = "denyChecker"
+
+	cfg = &aspect.CombinedConfig{
+		Aspect: &istioconfig.Aspect{
+			Kind:    name,
+			Adapter: "",
+			Inputs:  make(map[string]string),
+			Params:  new(structpb.Struct),
+		},
+		Builder: &istioconfig.Adapter{
+			Name:   "",
+			Kind:   "",
+			Impl:   name,
+			Params: new(structpb.Struct),
+		},
+	}
+)
+
+func TestNewPoolPanics(t *testing.T) {
+	sizeInt := -1 // this has type int; golint complains if we use a descriptive var declaration
+	defer func() {
+		if r := recover(); r == nil {
+			t.Error("Expected value in panic, got nothing.")
+		}
+	}()
+
+	newPool(uint(sizeInt))
+	t.Error("Expected panic in constructor due to size overflow")
+}
+
+func TestPoolSize(t *testing.T) {
+	blockChan := make(chan struct{})
+	testMngr := newTestManager(name, func() (*aspect.Output, error) {
+		<-blockChan
+		return &aspect.Output{Code: code.Code_OK}, nil
+	})
+	b := StaticBinding{
+		RegisterFn: func(r adapter.Registrar) error { return r.RegisterDenyChecker(testMngr) },
+		Manager:    testMngr,
+		Config:     cfg,
+		Methods:    []Method{Check},
+	}
+
+	// Easier than creating a new manager directly and having to register everything. We need all the config either way.
+	mgr, _ := NewMethodHandlers(1, b).(*methodHandlers)
+	underTest := newPool(1)
+
+	// Note: we allocate a result buffer of size two by passing in numAdapters = 2
+	res, enqueue := underTest.requestGroup(mgr.mngr, nil, nil, 2)
+
+	// Enqueue work which will not complete until blockChan is closed; since the pool size == 1 this blocks the queue
+	enqueue(context.Background(), cfg)
+
+	second := make(chan struct{})
+	go func() {
+		// this second enqueue will block until blockChan is closed
+		enqueue(context.Background(), cfg)
+		close(second)
+	}()
+
+	if len(res) != 0 {
+		t.Errorf("Expected result chan to be empty before work completes, got length: %d", len(res))
+	}
+
+	close(blockChan) // unblock the queue
+	<-second         // block the test thread till the second enqueue completes
+
+	// It takes a little time for the two goroutines to write their results; we loop to make the test more reliable.
+	for count := 0; len(res) != 2 && count < 5; count++ {
```

Comment:
can we do this without the sleeps?  some sort of loop on the res channel that pulls data into a slice with a deadline or something?